File: Minimal/src/Brontide.Minimal.Host/ProviderTrustOfflinePolicy.fs
namespace Brontide.Minimal.Host

open System

type ProviderTrustOfflineDecision =
    { Code: string
      MayContinueExistingService: bool
      MayStartProvider: bool
      Deadline: DateTimeOffset option
      RetryAt: DateTimeOffset option }

/// One host-owned availability decision. Grace preserves existing service only; it is not trust
/// evidence and never authorizes acquisition, launch, admission, or restart.
type ProviderTrustOfflinePolicy internal (grace: TimeSpan, retry: TimeSpan) =
    member _.Grace = grace
    member _.Retry = retry

    member _.Evaluate(
        now: DateTimeOffset,
        lastCurrent: DateTimeOffset option,
        pollCode: string,
        lastAttemptCode: string option,
        servingCount: int) =
        if String.IsNullOrWhiteSpace pollCode then
            invalidArg (nameof pollCode) "A poll outcome code is required."
        elif servingCount < 0 || servingCount > 64
             || (lastCurrent |> Option.exists (fun value ->
                    value > now || value.Ticks > DateTimeOffset.MaxValue.Ticks - grace.Ticks)) then
            { Code = "offline-observation-invalid"; MayContinueExistingService = false
              MayStartProvider = false; Deadline = None; RetryAt = None }
        else
            match lastCurrent with
            | None ->
                { Code = "offline-service-stop-required"; MayContinueExistingService = false
                  MayStartProvider = false; Deadline = None; RetryAt = None }
            | Some observed ->
                let deadline = observed.Add grace
                let eligible =
                    pollCode = "policy-poll-exhausted"
                    && (lastAttemptCode = Some "policy-distribution-transport-failed"
                        || lastAttemptCode = Some "policy-distribution-timeout")
                if not eligible then
                    { Code = "offline-service-stop-required"; MayContinueExistingService = false
                      MayStartProvider = false; Deadline = Some deadline; RetryAt = None }
                elif now >= deadline then
                    { Code = "offline-grace-expired"; MayContinueExistingService = false
                      MayStartProvider = false; Deadline = Some deadline; RetryAt = None }
                else
                    let remaining = deadline - now
                    let retryAt = if retry >= remaining then deadline else now.Add retry
                    { Code = if servingCount = 0 then "offline-idle" else "offline-existing-service"
                      MayContinueExistingService = servingCount > 0
                      MayStartProvider = false
                      Deadline = Some deadline
                      RetryAt = Some retryAt }

[<RequireQualifiedAccess>]
module ProviderTrustOfflinePolicy =
    let create (grace: TimeSpan) (retry: TimeSpan) =
        if grace <= TimeSpan.Zero || grace > TimeSpan.FromHours 24.0 then
            invalidArg (nameof grace) "Offline grace must be positive and no greater than twenty-four hours."
        if retry <= TimeSpan.Zero || retry > grace then
            invalidArg (nameof retry) "Retry must be positive and no greater than offline grace."
        ProviderTrustOfflinePolicy(grace, retry)

type ProviderTrustCadenceReconciliationVerdict =
    | NoEffectsConfirmed
    | EffectsAccountedFor
    | Unknown

type ProviderTrustCadenceReconciliationEvidence =
    { RunIdentity: ProviderTrustCadenceRunId
      AttemptIndex: int
      AttemptInstant: DateTimeOffset
      Verdict: ProviderTrustCadenceReconciliationVerdict }

/// Checks that host evidence names the exact interrupted attempt before selecting CBI48's durable
/// transition. The host owns the evidence; this module does not infer external effects.
[<RequireQualifiedAccess>]
module ProviderTrustCadenceReconciliation =
    let private current code snapshot: ProviderTrustCadenceJournalTransitionResult =
        { Code = code; Snapshot = snapshot }

    let private map acceptedCode (result: ProviderTrustCadenceJournalTransitionResult) =
        if result.Code = "durable-cadence-retry-ready"
           || result.Code = "durable-cadence-abandoned" then
            { result with Code = acceptedCode }
        else result

    let apply
        (journal: DurableProviderTrustCadenceJournal)
        (evidence: ProviderTrustCadenceReconciliationEvidence) =
        if isNull (box journal) then nullArg (nameof journal)
        if isNull (box evidence) then nullArg (nameof evidence)
        let snapshot = journal.Snapshot
        if snapshot.Phase <> "in-flight" then
            current "cadence-reconciliation-not-required" snapshot
        elif snapshot.RunIdentity <> evidence.RunIdentity
             || snapshot.NextCycleIndex <> evidence.AttemptIndex
             || snapshot.PreparedInstant <> evidence.AttemptInstant then
            current "cadence-reconciliation-mismatch" snapshot
        else
            match evidence.Verdict with
            | ProviderTrustCadenceReconciliationVerdict.Unknown ->
                current "cadence-reconciliation-deferred" snapshot
            | ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed ->
                journal.ResolveInterrupted ProviderTrustCadenceRecoveryDecision.Retry
                |> map "cadence-reconciliation-retry-ready"
            | ProviderTrustCadenceReconciliationVerdict.EffectsAccountedFor ->
                journal.ResolveInterrupted ProviderTrustCadenceRecoveryDecision.Abandon
                |> map "cadence-reconciliation-abandoned"

File: Reference/src/Brontide.Reference.Studio/ProviderTrustOfflinePolicy.cs
namespace Brontide.Reference.Studio;

public sealed record ProviderTrustOfflineDecision(
    string Code,
    bool MayContinueExistingService,
    bool MayStartProvider,
    DateTimeOffset? Deadline,
    DateTimeOffset? RetryAt);

/// <summary>
/// One host-owned availability decision. Grace preserves existing service only; it is not a trust
/// assertion and never authorizes acquisition, launch, admission, or restart.
/// </summary>
public sealed class ProviderTrustOfflinePolicy
{
    private ProviderTrustOfflinePolicy(TimeSpan grace, TimeSpan retry)
    {
        Grace = grace;
        Retry = retry;
    }

    public TimeSpan Grace { get; }
    public TimeSpan Retry { get; }

    public static ProviderTrustOfflinePolicy Create(TimeSpan grace, TimeSpan retry)
    {
        if (grace <= TimeSpan.Zero || grace > TimeSpan.FromHours(24))
            throw new ArgumentOutOfRangeException(nameof(grace));
        if (retry <= TimeSpan.Zero || retry > grace)
            throw new ArgumentOutOfRangeException(nameof(retry));
        return new(grace, retry);
    }

    public ProviderTrustOfflineDecision Evaluate(
        DateTimeOffset now,
        DateTimeOffset? lastCurrent,
        string pollCode,
        string? lastAttemptCode,
        int servingCount)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(pollCode);
        if (servingCount is < 0 or > 64 || lastCurrent > now
            || lastCurrent is { } observed
                && observed.Ticks > DateTimeOffset.MaxValue.Ticks - Grace.Ticks)
            return new("offline-observation-invalid", false, false, null, null);

        if (lastCurrent is null)
            return new("offline-service-stop-required", false, false, null, null);
        var deadline = lastCurrent.Value.Add(Grace);

        var eligible = pollCode == "policy-poll-exhausted"
            && lastAttemptCode is "policy-distribution-transport-failed"
                or "policy-distribution-timeout";
        if (!eligible)
            return new("offline-service-stop-required", false, false, deadline, null);
        if (now >= deadline)
            return new("offline-grace-expired", false, false, deadline, null);

        var remaining = deadline - now;
        var retryAt = Retry >= remaining ? deadline : now.Add(Retry);
        return servingCount == 0
            ? new("offline-idle", false, false, deadline, retryAt)
            : new("offline-existing-service", true, false, deadline, retryAt);
    }
}

public enum ProviderTrustCadenceReconciliationVerdict
{
    NoEffectsConfirmed,
    EffectsAccountedFor,
    Unknown,
}

public sealed record ProviderTrustCadenceReconciliationEvidence(
    ProviderTrustCadenceRunId RunIdentity,
    int AttemptIndex,
    DateTimeOffset AttemptInstant,
    ProviderTrustCadenceReconciliationVerdict Verdict);

/// <summary>
/// Checks that host evidence names the exact interrupted attempt before selecting CBI48's durable
/// transition. The host owns the evidence; this class does not infer external effects.
/// </summary>
public static class ProviderTrustCadenceReconciliation
{
    public static ProviderTrustCadenceJournalTransitionResult Apply(
        DurableProviderTrustCadenceJournal journal,
        ProviderTrustCadenceReconciliationEvidence evidence)
    {
        ArgumentNullException.ThrowIfNull(journal);
        ArgumentNullException.ThrowIfNull(evidence);

        var snapshot = journal.Snapshot;
        if (snapshot.Phase != "in-flight")
            return new("cadence-reconciliation-not-required", snapshot);
        // One verdict for everything the cycle did was right while a cycle was one loop. A governed
        // attempt recorded the cursor it was about to act on, and two of the three things it can do
        // are derivable from that, so this verdict would be speaking for effects the host need not
        // have inspected.
        if (snapshot.Cursor is not null)
            return new("cadence-reconciliation-governed", snapshot);
        if (snapshot.RunIdentity != evidence.RunIdentity
            || snapshot.NextCycleIndex != evidence.AttemptIndex
            || snapshot.PreparedInstant != evidence.AttemptInstant)
            return new("cadence-reconciliation-mismatch", snapshot);

        return evidence.Verdict switch
        {
            ProviderTrustCadenceReconciliationVerdict.Unknown =>
                new("cadence-reconciliation-deferred", snapshot),
            ProviderTrustCadenceReconciliationVerdict.NoEffectsConfirmed =>
                Map(journal.ResolveInterrupted(ProviderTrustCadenceRecoveryDecision.Retry),
                    "cadence-reconciliation-retry-ready"),
            ProviderTrustCadenceReconciliationVerdict.EffectsAccountedFor =>
                Map(journal.ResolveInterrupted(ProviderTrustCadenceRecoveryDecision.Abandon),
                    "cadence-reconciliation-abandoned"),
            _ => throw new ArgumentOutOfRangeException(nameof(evidence)),
        };
    }

    private static ProviderTrustCadenceJournalTransitionResult Map(
        ProviderTrustCadenceJournalTransitionResult result,
        string acceptedCode) =>
        result.Code is "durable-cadence-retry-ready" or "durable-cadence-abandoned"
            ? new(acceptedCode, result.Snapshot)
            : new(result.Code, result.Snapshot);
}
